cEP-0020.md

```diff
@@ -0,0 +1,590 @@
+
+A mockup of a basic `Activity` Django model would look something
+like:
+
+```python
+class Actvity(models.Model):
+    name = models.CharField(max_length=500)
+    details = models.TextField()
+    points = models.IntegerField()
+    number_of_times = model.IntegerField(default=0)
+```
+
+`number_of_times` field store the value of number of times
+an activity has been done by the same user.
+
+The `User` model will have a ManyToMany relationship with
+the `Activity` model with a method named
+`count_similar_activity` which will check if the action
+by the user has been done previously? If Yes, it will
+increase `number_of_times` field with one.
+
+#### Creating the Level Model
+
+A mockup of basic `Level` Django model would look something like:
+
+```python
+class Level(models.Model):
+    name = models.CharField()
+    points_required = model.IntegerField()
+    user = models.ManytoManyField(User)
+    privileges = models.ForeignKey(Privilege)
+
+
+class Privilege(models.Model):
+    name = models.CharField(max_length=500)
+```
+
+And a query to create a level would be:
+
+```python
+privilege = Privilege.objects.create(
+    name='Get assigned to a newcomer issue'
+    )
+level1 = Level.obejects.create(
+    name='Newbie',
+    points_required=5,
+    privileges=privilege
+    )
+```
+
+#### Extending the User Model to add points and Levels
+
+Whenever a user perform an activity his points will be increased.
+So, there should be some methods to check if the user has achieved
+a new level:
+
+```python
+class User(models.Model):
+    ...
+    ...
+    points = models.IntegerField()
+    level = models.CharField()
+    ...
+
+    def add_points()
+        a_method_to_update_points_and_level()
+        a_method_for_the_log_of_the_activities()
+```
+
+#### Creating the Badges model
+
+```python
+class Badge(models.Model):
+    name = models.CharField()
+    description = model.TextField()
+```
+
+The Badge model much like our Level model, defines a range of
+possible awards that a user can receive. Unlike levels, users
+can have many badges at one time. This sets up a many-to-many
+relationship between badges and users, so there is a need of an
+intermediate model to track which badges a user has earned.
+The model will be called `Achievement`:
+
+```python
+class Achievement(models.Model):
+    user_id = ManyToManyField(User)
+    bage_id = ManyToManyField(Badge)
+```
+
+Now, to make awarding badges easier, let’s also add a
+convenience method to the User model:
+
+```python
+def award_badge(name):
+    badge = Badge.objects.filter(name=name)
+    badges.award(name)
+```
+
+### Deployment and The leaderboard
+
+There will be Django management commands for each of the
+needed processes which will be triggered on each build of
+netlify and show the useful updated user metrics and stats
+on the leaderboard.
+
+The leaderboard will have the following column:
+
+| Username | Important Activity Metrics | Points | Level and Privileges
```

Comment:
shorten headers to get onto a single line.
splitting onto two lines is invalid syntax

at least "Level and Privileges" -> "Levels"
as Privileges are implicitly the same thing.